### PR TITLE
test: add tests for incremental compilation

### LINKED ISSUE #2968: Write unit tests for incremental compilation
We need to write unit tests for incremental compilation. I suggest something like:

```scala
val flix = new Flix()
flix.addSourceCode("test.flix", 
"""
pub def f(x: Int32, y: Int32): Int32 = x + y

pub def g(): Int32 = f(1, 2)
""")
flix.compile() // assert that it works.
flix.addSourceCode("test.flix", 
"""
pub def f(x: String, y: String): Int32 = x + y

pub def g(): String= f("hello", "world")
""")
flix.compile() // assert that it works.
```

The point is to add a small program, make incompatible changes, and then recompile it.

I suggest to start a PR with just one test case so we can discuss the structure.

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #2968

d08eebf was committed before adding a gpg key, so it's okay even though it's unverified

### CODE REVIEW

**Comment by jaschdoc** [comment]
@magnus-madsen what do you think of it so far?

**Comment by magnus-madsen** [comment]
lets ask @mlutze too what he thinks :)

File: main/test/ca/uwaterloo/flix/language/phase/TestIncremental.scala [hunk 1]
```diff
@@ -86,11 +111,138 @@ class TestIncremental extends FunSuite with BeforeAndAfter {
...
+
+
+  /**
+    * Helper function that handles compilation errors.
+    *
+    * @param res compilation result.
+    * @tparam T type of success value.
+    * @tparam E type of failure value.
+    */
+  private def expectFailure[T, E](res: => Validation[T, E]): Unit = res match {
```
Comment (mlutze) [inline main/test/ca/uwaterloo/flix/language/phase/TestIncremental.scala:244]: I think we can put this in the TestUtils (or wherever `expectError` is)

**Review by mlutze — COMMENTED** [review]
I think there are at least two other subjects to test:
* Superclass relationship
* Type aliases

**Comment by jaschdoc** [comment]
@magnus-madsen @mlutze Ready for review

File: main/test/ca/uwaterloo/flix/language/phase/TestIncremental.scala [hunk 2]
```diff
@@ -86,11 +135,181 @@ class TestIncremental extends FunSuite with BeforeAndAfter {
...
+         |    pub def cda(l: L[a]): a = l.x
+         |}
+         |""".stripMargin)
+    flix.addSourceCode(FileF,
+      s"""
+         |pub type alias L[a] = { x :: a }
+         |""".stripMargin)
+
+    flix.compile().get
+  }
```
Comment (mlutze) [inline main/test/ca/uwaterloo/flix/language/phase/TestIncremental.scala:291]: newline

**Comment by jaschdoc** [comment]
> Looks like everything's covered. My only complaint is that it's not clear to me, for each test, what the focus of the test is, since a lot of them contain several sources. I'm wondering if it might be nicer to have one small test for each declaration, and then several "big jumble" tests. What do you think?

Yes this is definitely something I struggled with! I was wondering how tightly coupled the Flix program should be so I just sort of put everything into the `C` type class which then has the nasty side effect that existing tests sometimes have to be changed. Maybe `C` should be split out into other files so tests are more isolated and then have bigger integration / "big jumble" tests like you suggest.

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**